### PR TITLE
Added docker-compose.yml with excluded Memoryleakfix.

### PR DESCRIPTION
**Pull request by bashbers**
See https://github.com/TheStaticVoid/StaTech-Industry/issues/163 for more information as of why this is excluded.

### CODE REVIEW

**Comment by bashbers** [comment]
@TheStaticVoid I tried using the server files as input for the CF_PAGE_URL, but that did not work out as intended. See https://github.com/itzg/docker-minecraft-server/issues/2007#issuecomment-1470550728 for more information on why the docker image author is suggesting to use the client zip in combination with an excluded modlist.

**Comment by TacoMonkey11** [comment]
Does this deal with mods which have opted out from third-party downloads? Or is that still a manual responsibility for the user?

**Comment by bashbers** [comment]
Yes, we are using a Curseforge developer API key.

**Comment by TacoMonkey11** [comment]
```
minecraft  | [mc-image-helper] 00:55:26.025 WARN  : The authors of the mod 'Immersive structures' have disallowed project distribution. Manually download the file '[Universal]Immersive Structures-2.0.7a' from https://www.curseforge.com/minecraft/mc-mods/immersive-structures and supply via downloads repo or separately.
```

I don't believe the API key allows third-parties to download those mods which don't allow it. It is still up to the user

**Comment by TheStaticVoid** [comment]
The API key should allow it. It's blocked on launchers like Prism because we don't have one set.

**Comment by TheStaticVoid** [comment]
Just ran through the process and also experienced the warning message taco experienced. Not going to be able to merge this unless we find a way to download those as well.

**Comment by bashbers** [comment]
Ah I did not know that, that sucks.. These mods are just opting out of 3rd-party downloads and there's nothing we can do about that?

**Comment by TheStaticVoid** [comment]
Not within the CurseForge API

**Comment by bashbers** [comment]
Looked further into it and came across https://github.com/itzg/docker-minecraft-server/issues/2174

When this is merged into master, then I'll look into making an init-container to download the thirdparty mods! :)

**Comment by TheStaticVoid** [comment]
I'm not merging to master unless we are able to get those third-party mods as well. If we include as is it would just be more confusing to the user.

**Comment by bashbers** [comment]
Ah my previous comment might have been confusing. I was referring to the PR I linked, not this PR. The PR I linked is able to solve the third-party mods problem using an init-container, but since it's not merged yet, I would like to wait for any further developments on the PR.

**Comment by TheStaticVoid** [comment]
Would it be better to download server files from a github release?

**Comment by TacoMonkey11** [comment]
> The latest file will be located and used by default, but if a specific version is desired you can use one of the following options. With any of these options do not select a server file -- they lack the required manifest and defeat the ability to consistently automate startup.

https://docker-minecraft-server.readthedocs.io/en/latest/mods-and-plugins/curseforge/#usage

**Comment by bashbers** [comment]
The downloaded server files could work aswell, will look into it.

**Comment by TacoMonkey11** [comment]
I reviewed the issue linked in this discussion about init-containers, so I have decided to create one myself, and I have it working rather well.

new contents of docker-compose.yml would be as follows:

```yaml
version: '3'

services:
  cf-initcontainer:
    image: tacomonkey/cf-initcontainer:latest
    container_name: cf-initcontainer
    volumes:
      - "./downloads:/downloads"
    environment:
      - CF_API_KEY=${CF_API_KEY}
      - MODPACK_ID=863689 # project id of the modpack
      - FILE_ID=4566222 #file id of the specific modpack version

  minecraft:
    image: itzg/minecraft-server:java17-graalvm-ce
    container_name: minecraft
    ports:
      - "25565:25565"
    volumes:
      - "./mc:/data"
      - "./downloads:/downloads"
    environment:
      - EULA=TRUE
      - MEMORY=8G
      - TYPE=AUTO_CURSEFORGE
      - CF_PAGE_URL=https://www.curseforge.com/minecraft/modpacks/statech-industry
      - CF_API_KEY=${CF_API_KEY}
      - CF_EXCLUDE_MODS=627566 # Disabled Memoryleakfix mod, with this mod the server wouldnt boot. See https://github.com/TheStaticVoid/StaTech-Industry/issues/163
    depends_on:
      cf-initcontainer:
        condition: service_completed_successfully
    tty: true
    stdin_open: true
    restart: always
```

The init-container will start a script which uses the CF API to figure out which mods need to be downloaded, and then forges a download link to use alongside the CF_DOWNLOADS_REPO option in docker-minecraft-server. This may or may not be against curseforge TOS, that is up to interpretation.

**Comment by bashbers** [comment]
Nicely done, one thing I don't like it the inclusion of your own image. I much rather rely on something generic like a curlcontainer used in the PR I linked.

**Comment by TacoMonkey11** [comment]
It seems to be a network issue. If the curl command to query the CF API wasn't silent, it probably would result in an error. I don't actually know what causes this issue, but I believe it may be solved by specifying a different DNS like `8.8.8.8`. I had similar issues on windows.

**Comment by bashbers** [comment]
This was on a Ubuntu VPS. I think we should make the curl more verbose so we can see more clearly what's going on?

**Comment by TacoMonkey11** [comment]
That's a good idea, I can just remove the silent flag from curl.

**Comment by TheStaticVoid** [comment]
Do we need the FILE_ID now then? Would require updating after a patch to update the id each time.

**Comment by bashbers** [comment]
The FILE_ID is needed unless we implement a way to get the latest FILE_ID from CF. I would suggest to leave it as is and just update the id once a new version is released. This way people have control over their deployed server version instead of having to rely on the latest version.

**Comment by TacoMonkey11** [comment]
@bashbers could you try to run the container again? I've updated the image to now use debian-slim instead of alpine which seems to have fixed the network inconsistencies. The only downside is that the image is now ~47MB instead of 6MB

**Comment by bashbers** [comment]
image
docker inspect results:

```sh
xxx@xxx:~/minecraft/test# docker inspect tacomonkey/cf-initcontainer
[
    {
        "Id": "sha256:5c8f84ef3c5c380955ebdff8b1f46c3a5821095781783b1ed3e4342fe1f6f11d",
        "RepoTags": [
            "tacomonkey/cf-initcontainer:latest"
        ],
        "RepoDigests": [
            "tacomonkey/cf-initcontainer@sha256:473b1327faf18fea6434a61e07c1827d98df6056b8e912883f511cf95c66431b"
        ],
        "Parent": "",
        "Comment": "buildkit.dockerfile.v0",
        "Created": "2023-06-05T21:05:38.0124667Z",
        "Container": "",
        "ContainerConfig": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "DockerVersion": "",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": null,
            "Image": "",
            "Volumes": {
                "/downloads": {}
            },
            "WorkingDir": "/downloads",
            "Entrypoint": [
                "/bin/bash",
                "/usr/src/find-denied-mods.sh"
            ],
            "OnBuild": null,
            "Labels": null
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 105939178,
        "VirtualSize": 105939178,
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/731ed748fbfd63f0390473414047326af7716d4ab5063dcc06d53acb044fc6a1/diff:/var/lib/docker/overlay2/ce109c099c810abcdfafcfafa890d9a3f27fd5ea12a62975023d05d923e11fe7/diff:/var/lib/docker/overlay2/4c26a94e73aeec5df433fc683d623989c9c81465ce2e065d3c7b66c912739f38/diff",
                "MergedDir": "/var/lib/docker/overlay2/ffee31666f36429faef971cb7852b7544cf8b96131bbc26b4d4ea0f195f2e01d/merged",
                "UpperDir": "/var/lib/docker/overlay2/ffee31666f36429faef971cb7852b7544cf8b96131bbc26b4d4ea0f195f2e01d/diff",
                "WorkDir": "/var/lib/docker/overlay2/ffee31666f36429faef971cb7852b7544cf8b96131bbc26b4d4ea0f195f2e01d/work"
            },
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:8e14b8d986dc16da49658ff2abc8e0f0d2760d0e215b385b35e0837da1502bdf",
                "sha256:57ccbfe0778a91fc6a81553838aa2c9b035f4e11adba825b643b366cedb238c5",
                "sha256:d03af2f65d3e00548f18ff6c89d772ee072383c8c4254fed04f262799a314639",
                "sha256:09a08584c6688778bbd222dcba68eec33e3a8cec21cd02b548c58b144f3979da"
            ]
        },
        "Metadata": {
            "LastTagTime": "0001-01-01T00:00:00Z"
        }
    }
]
```

**Comment by bashbers** [comment]
now the script is crashing on the `mv` commands:

image

**Comment by TacoMonkey11** [comment]
This seems rather simple to fix, as everything actually worked correctly, I just need to ignore that "error" by adding `--force` to the `mv` command

File: README.md [hunk 1]
```diff
@@ -22,3 +22,55 @@ I recommend the use of [Prism Launcher](https://prismlauncher.org/) for handling
 3. You will be prompted with a menu describing that some mods are blocked from download on third-party launchers. Add the folder you use as your downloads folder using `Add Download Folder` button, then click `Open Missing` to open the mod downloads. Save them to your downloads and they will be detected by the launcher.
 4. Click "Ok" and allow for the process to finish. 
 5. Click "Edit" under the StaTech Industry instance and ensure you have at least 4-6 GB or RAM dedicated to the pack, as well as your version of Java is Java 17.
+
+
+## Server setup using Docker-compose
```
Comment (TheStaticVoid) [inline README.md:27]: Keep this section in the README.md. It should link to the Server Setup wiki page where we will include the rest of the sections from this README

File: docker-compose.yml
Comment (TheStaticVoid) [inline docker-compose.yml:1]: Update this compose with the latest revision from the discussion

File: .env.example
Comment (TheStaticVoid) [inline .env.example:1]: Did we settle on needing to double the $ signs or no?

**Review by TheStaticVoid — CHANGES_REQUESTED** [review]
Now that we have something that seems usable let's get the ball rolling.

File: .env.example
Comment (TheStaticVoid) [reply .env.example:1]: I tested using double $ as well. Would need confirmation from @bashbers

Comment (bashbers) [reply .env.example:1]: Its really weird, in the cf minecraft container it needs the double `$` but your container does not @TacoMonkey11 .

Comment (TacoMonkey11) [reply .env.example:1]: I believe that another issue may be a play then since the API key should not be that inconsistent between users

**Comment by TacoMonkey11** [comment]
Since this seems rather stale, I think I will make these changes myself alongside ignoring removing all the other clientside mods.

I'll make a new PR

**Pull request closed without merging**